ts/src/registry/index.ts

```diff
@@ -0,0 +1,231 @@
+/*
+	Copyright 2022 Loophole Labs
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		   http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+import {ScaleFunc} from "@loopholelabs/scalefile"
+import { sha256 } from "js-sha256";
+import * as fs from "fs";
+import { webcrypto as crypto } from "crypto";
+import https from "https";
+
+export type PullPolicy = string;
+
+export const AlwaysPullPolicy: PullPolicy = "always";
+export const IfNotPresentPullPolicy: PullPolicy = "if-not-present";
+export const NeverPullPolicy: PullPolicy = "never";
+
+export var ErrHasMismatch = new Error("hash mismatch");
+export var ErrNoFunction = new Error("function does not exist and pull policy is never");
+
+export interface Config {
+  pullPolicy: PullPolicy
+  cacheDirectory: string
+  apiKey: string,
+  apiBaseUrl: string,
+  organization: string,
+}
+
+export type Option = (c: Config) => void;
+
+export function WithPullPolicy(pullPolicy: PullPolicy): Option {
+  return (c: Config) => {
+    c.pullPolicy = pullPolicy;
+  }
+}
+
+export function WithCacheDirectory(cacheDirectory: string): Option {
+  return (c: Config) => {
+    c.cacheDirectory = cacheDirectory;
+  }
+}
+
+export function WithApiKey(apiKey: string): Option {
+  return (c: Config) => {
+    c.apiKey = apiKey;
+  }
+}
+
+export function WithApiBaseUrl(apiBaseUrl: string): Option {
+    return (c: Config) => {
+        c.apiBaseUrl = apiBaseUrl;
+    }
+}
+
+export function WithOrganization(organization: string): Option {
+    return (c: Config) => {
+        c.organization = organization;
+    }
+}
+
+export async function New(func: string, tag: string, ...opts: Option[]): Promise<ScaleFunc> {
+  const config: Config = {
+    pullPolicy: AlwaysPullPolicy,
+    cacheDirectory: "~/.cache/scale/functions",
```

Comment:
Some way to use the homedir lib here too? The `~` won't work on some operating systems.